Implement single axis zoom

lucasgruwez opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please describe the pull request, using the questions below as guidance, and link to any relevant issues and PRs:

- Why is this change necessary?
- What problem does it solve?
- What is the reasoning for this implementation?

Additionally, please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".

If possible, please provide a minimum self-contained example.  If you have used
generative AI as an aid in preparing this PR, see

https://matplotlib.org/devdocs/devel/contribute.html#restrictions-on-generative-ai-usage
-->

Tangent to #30541 
Allows for zooming in a single axis using the zoom rectangle by dragging only in one direction.

https://github.com/user-attachments/assets/28228153-5231-4228-9947-e7d520238cb2

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

rcomer commented:
I note that the existing x-/y- modifiers do not really work when using my laptop's touchpad[^1], but this proposal does work nicely :+1: 

[^1]: If I hold `x` and the left button, nothing happens when I drag.  I can get it to work by making the box and then pressing `x` before I let go, but I don't get the visual of the box turning into two parallel lines.  If I use an external mouse, everything works as it should so it definitely seems to be an issue with the touchpad.

timhoffm commented:
The range marker is a good visualization to indicate the difference. Though, I would like to additionally keep dashed lines as an extension, because that helps to ensure relevant data points are included.

That said, all the visual changes would need refactoring and backend-dependent implementation as currently all visualization is done via `RubberBandBase` and that does only cover the rectangle without the ability for further modification.
https://github.com/matplotlib/matplotlib/blob/5b38f50ae200e09419f929a5f3760ea5ae099d2e/lib/matplotlib/backend_tools.py#L327

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2935,6 +2935,17 @@ def draw_rubberband(self, event, x0, y0, x1, y1):
+        Draw line with whiskers to indicate single axis zoom
+
+        We expect that ``x0 == x1`` or ``y0 == y1``. Else nothing will draw
+        `ws` is the whisker size in pixels.
```
```suggestion
        *ws* is the whisker size in pixels.
```
We mark parameters this way. Single ticks will try to link somewhere and induce a failure in the doc build.

lucasgruwez commented:
All backends have been updated. As previously mentioned, I was unable to test gtk on my machine, but I have tested all other backends. Several other fixes were also included:
- Ensure that the zoom rubber band is dashed in all backends (it was not in WebAgg).
- Fix the Tk rubber band: the implicit black line colour was not working, so it was set explicitly instead.

timhoffm commented:
> All backends have been updated. As previously mentioned, I was unable to test gtk on my machine, but I have tested all other backends. Several other fixes were also included:
> 
> * Ensure that the zoom rubber band is dashed in all backends (it was not in WebAgg).
> * Fix the Tk rubber band: the implicit black line colour was not working, so it was set explicitly instead.

Could you move them out into a separate PR? It makes reviewing this PR simpler if we don't have multiple topics.

This PR is a good prove-of-concept and from a first glance feature complete - thanks for implementing all the backends!

We still need to think about the architecture and how to bring it into the main library. As is, the `draw_whisker` method is an API change that will break third-party backends. We'll have to decide

- API design: whether that method is the right approach (other alternatives could be integrate it into the rubberband, or start a more general approach for drawing some primitives in the backend in the anticipation that we will have more interactivity (e.g. #30516).
- API compatibility: As is this is breaking 3rd party backends, which is a no-go. We could either work around this through special-casing (test if the method is available in the backend) or go for a more structured approach of API versioning #30559.

lucasgruwez commented:
Tested with `mplcairo` third party backend and got the following results
- `mplcairo.macosx`: Whiskers inherited from mpl macosx class
- `mplcairo.qt`: No whiskers drawn, box only. No errors thrown
- `mplcairo.wx`: Scaling is off on my machine, whiskers still work though
- `mplcairo.tk`: Whiskers inherited from mpl tk class
- `mplcairo.gtk`: Whiskers inherited from mpl gtk class

lucasgruwez commented:
Having a more thorough read of the issues you've linked, I think we can consider the following implementation:

a `draw_line()` method be implemented for each backend, taking in color/stroke/dash parameters. `draw_rubberband()` and `draw_whiskers()` could then make use of this method in a backend-agnostic way.

I see that this could have some ramifications on third party backends, as it could break `draw_rubberband()` as they won't have a `draw_line()` method (unless subclassed from a mpl backend). Somewhat uncertain about the versioning system, is this for third party backends only? or is there concern that mpl native backends could misalign with the package itself?

timhoffm commented:
> Somewhat uncertain about the versioning system, is this for third party backends only? or is there concern that mpl native backends could misalign with the package itself?

native backends could partially misalign, i.e. an intenal backend doesn't have to be on the newest version of the API, but of course it must have a version that is supported, i.e. if at some point matplotlib supports backend versions v1 and v2, then every internal backend must be one of these versions. For example it would be permissible if we don't implement a new API for all backends right away.

timhoffm commented:
> see that this could have some ramifications on third party backends, as it could break `draw_rubberband()` as they won't have a `draw_line()` method (unless subclassed from a mpl backend).

For the time being, I would leave `draw_rubberband()` it's working as is. We can refactor that later when/if backends have gained more general drawing capability.

lucasgruwez commented:
After an almost year-long nap, I’ve revived this PR 😅

I updated the branch to current `main` and addressed the backend compatibility concerns from the earlier discussion:

- `draw_whiskers()` and `remove_whiskers()` are optional no-op hooks in `NavigationToolbar2`, so third-party backends continue to work without displaying the indicator.
- WebAgg clients advertise whisker support before receiving whisker events. Older clients such as ipympl do not receive an event they do not understand.
- The proposed optional hooks are documented as backend API 1.2.
- Whiskers use a thin black stroke with a white outline, keeping them visible on light and dark plots.
- I fixed whisker cleanup on canceled zooms, macOS redraw trails, and WebAgg flicker.

The unrelated rubberband fixes were split into #30560, which has since merged.

I tested the updated implementation across macOS, wx, Tk, Qt, GTK3, and WebAgg in light and dark mode (see screenshots)

<img width="2040" height="1200" alt="13241" src="https://github.com/user-attachments/assets/2378e3ea-c686-4ab3-9011-1e924f527f88" />

Should backends without whisker support still get inferred single-axis zoom? They currently do because the gesture logic lives in `NavigationToolbar2`. If the visual indicator should be required, I can add an explicit `supports_single_axis_zoom = False` capability and enable it only in backends that implement whiskers. That would preserve the existing rectangle-zoom behavior for older backends.

Since #30953 now covers a broader overlay architecture, are you comfortable merging these optional backend hooks as an incremental step, with a later migration to the overlay system? I’m happy to adjust the direction before doing more cleanup.

lucasgruwez commented:
Light mode behvior for reference, relatively unchanged from previous implementation

<img width="2040" height="1200" alt="41149" src="https://github.com/user-attachments/assets/b7304058-b31f-4fda-86fc-21a69f8215d8" />
